{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2ff480c5-643f-4b70-88ef-578d510f8547",
   "metadata": {},
   "source": [
    "### Importing Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0068f02b-544b-43de-bef6-36a9b98ebb5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import List\n",
    "import os\n",
    "import pdb\n",
    "import re\n",
    "import ast\n",
    "import json\n",
    "import random\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from openai import OpenAI\n",
    "from importlib import resources\n",
    "from IPython.display import display, clear_output\n",
    "from chunking_evaluation.utils import rigorous_document_search\n",
    "from chunking_evaluation.evaluation_framework.base_evaluation import BaseEvaluation\n",
    "\n",
    "class SyntheticEvaluation(BaseEvaluation):\n",
    "    def __init__(self, corpora_paths: List[str], queries_csv_path: str, chroma_db_path: str = None,\n",
    "                 openai_api_key=None):\n",
    "        super().__init__(questions_csv_path=queries_csv_path, chroma_db_path=chroma_db_path)\n",
    "        self.corpora_paths = corpora_paths\n",
    "        self.questions_csv_path = queries_csv_path\n",
    "        self.client = OpenAI(base_url=\"Your-URL-HERE\", api_key=\"XXXX-Your-API-Key-Here-XXX\")\n",
    "        self.synth_questions_df = None\n",
    "\n",
    "        with resources.as_file(resources.files('chunking_evaluation.evaluation_framework') / 'prompts') as prompt_path:\n",
    "            with open(os.path.join(prompt_path, 'question_maker_system.txt'), 'r') as f:\n",
    "                self.question_maker_system_prompt = f.read()\n",
    "\n",
    "            with open(os.path.join(prompt_path, 'question_maker_approx_system.txt'), 'r') as f:\n",
    "                self.question_maker_approx_system_prompt = f.read()\n",
    "\n",
    "            with open(os.path.join(prompt_path, 'question_maker_user.txt'), 'r') as f:\n",
    "                self.question_maker_user_prompt = f.read()\n",
    "\n",
    "            with open(os.path.join(prompt_path, 'question_maker_approx_user.txt'), 'r') as f:\n",
    "                self.question_maker_approx_user_prompt = f.read()\n",
    "\n",
    "    def _save_questions_df(self):\n",
    "        self.synth_questions_df.to_csv(self.questions_csv_path, index=False)\n",
    "\n",
    "    def _tag_text(self, text):\n",
    "        chunk_length = 100\n",
    "        chunks = []\n",
    "        tag_indexes = [0]\n",
    "        start = 0\n",
    "        while start < len(text):\n",
    "            end = start + chunk_length\n",
    "            chunk = text[start:end]\n",
    "            if end < len(text):\n",
    "                # Find the last space within the chunk to avoid splitting a word\n",
    "                space_index = chunk.rfind(' ')\n",
    "                if space_index != -1:\n",
    "                    end = start + space_index + 1  # Include the space in the chunk\n",
    "                    chunk = text[start:end]\n",
    "            chunks.append(chunk)\n",
    "            tag_indexes.append(end)\n",
    "            start = end  # Move start to end to continue splitting\n",
    "\n",
    "        tagged_text = \"\"\n",
    "        for i, chunk in enumerate(chunks):\n",
    "            tagged_text += f\"<start_chunk_{i}>\" + chunk + f\"<end_chunk_{i}>\"\n",
    "\n",
    "        return tagged_text, tag_indexes\n",
    "\n",
    "\n",
    "    def _extract_question_and_approx_references(self, corpus, document_length=4000, prev_questions=[]):\n",
    "        \n",
    "        if len(corpus) > document_length:\n",
    "            start_index = random.randint(0, len(corpus) - document_length)\n",
    "            document = corpus[start_index: start_index + document_length]\n",
    "        else:\n",
    "            start_index = 0\n",
    "            document = corpus\n",
    "\n",
    "        if prev_questions is not None:\n",
    "            if len(prev_questions) > 20:\n",
    "                questions_sample = random.sample(prev_questions, 20)\n",
    "                prev_questions_str = '\\n'.join(questions_sample)\n",
    "            else:\n",
    "                prev_questions_str = '\\n'.join(prev_questions)\n",
    "        else:\n",
    "            prev_questions_str = \"\"\n",
    "\n",
    "        tagged_text, tag_indexes = self._tag_text(document)\n",
    "\n",
    "        completion = self.client.chat.completions.create(\n",
    "            model=\"microsoft/phi-4\",\n",
    "            messages=[\n",
    "                {\"role\": \"system\", \"content\": self.question_maker_approx_system_prompt},\n",
    "                {\"role\": \"user\",\n",
    "                 \"content\": self.question_maker_approx_user_prompt.replace(\"{document}\", tagged_text).replace(\n",
    "                     \"{prev_questions_str}\", prev_questions_str)}\n",
    "            ]\n",
    "        )\n",
    "\n",
    "        response = completion.choices[0].message.content \n",
    "\n",
    "        try:\n",
    "            # Step 1: Remove the ```json\\n prefix and ``` suffix\n",
    "            json_str = response.strip().split('```json\\n', 1)[-1].rsplit('\\n```', 1)[0]\n",
    "            json_response = ast.literal_eval(json_str)\n",
    "            print(json_response)\n",
    "        except Exception as e:\n",
    "            print(e)\n",
    "            raise ValueError(\"The response is not in expected format \")\n",
    "            \n",
    "        \n",
    "        try:\n",
    "            text_references = json_response['references']\n",
    "        except KeyError:\n",
    "            raise ValueError(\"The response does not contain a 'references' field.\")\n",
    "\n",
    "        try:\n",
    "            question = json_response['question']\n",
    "        except KeyError:\n",
    "            raise ValueError(\"The response does not contain a 'question' field.\")\n",
    "\n",
    "        \n",
    "        references = []\n",
    "    \n",
    "        for reference in text_references:\n",
    "            try:\n",
    "                reference_keys = list(reference.keys())\n",
    "    \n",
    "                if len(reference_keys) != 3:\n",
    "                    raise ValueError(\n",
    "                        f\"Each reference must have exactly 3 keys: 'content', 'start_chunk', and 'end_chunk'. Got keys: {reference_keys}\")\n",
    "    \n",
    "                if 'start_chunk' not in reference_keys or 'end_chunk' not in reference_keys:\n",
    "                    raise ValueError(\"Each reference must contain 'start_chunk' and 'end_chunk' keys.\")\n",
    "    \n",
    "                if 'end_chunk' not in reference_keys:\n",
    "                    reference_keys.remove('content')\n",
    "                    reference_keys.remove('start_chunk')\n",
    "                    end_chunk_key = reference_keys[0]\n",
    "                    end_index = start_index + tag_indexes[reference[end_chunk_key] + 1]\n",
    "                else:\n",
    "                    end_index = start_index + tag_indexes[reference['end_chunk'] + 1]\n",
    "    \n",
    "                start_index = start_index + tag_indexes[reference['start_chunk']]\n",
    "                references.append((corpus[start_index:end_index], start_index, end_index))\n",
    "                \n",
    "            except Exception as e:\n",
    "                print(e)\n",
    "                \n",
    "        return question, references\n",
    "\n",
    "    def _extract_question_and_references(self, corpus, document_length=4000, prev_questions=[]):\n",
    "        if len(corpus) > document_length:\n",
    "            start_index = random.randint(0, len(corpus) - document_length)\n",
    "            document = corpus[start_index: start_index + document_length]\n",
    "        else:\n",
    "            document = corpus\n",
    "\n",
    "        if prev_questions is not None:\n",
    "            if len(prev_questions) > 20:\n",
    "                questions_sample = random.sample(prev_questions, 20)\n",
    "                prev_questions_str = '\\n'.join(questions_sample)\n",
    "            else:\n",
    "                prev_questions_str = '\\n'.join(prev_questions)\n",
    "        else:\n",
    "            prev_questions_str = \"\"\n",
    "\n",
    "        completion = self.client.chat.completions.create(\n",
    "            model=\"microsoft/phi-4\",\n",
    "            messages=[\n",
    "                {\"role\": \"system\", \"content\": self.question_maker_system_prompt},\n",
    "                {\"role\": \"user\", \"content\": self.question_maker_user_prompt.replace(\"{document}\", document).replace(\n",
    "                    \"{prev_questions_str}\", prev_questions_str)}\n",
    "            ]\n",
    "        )\n",
    "\n",
    "        response = completion.choices[0].message\n",
    "        json_response = json.loads(response.json())\n",
    "        \n",
    "        input_str = json_response['content']\n",
    "        # new_string = self.remove_think_tags(input_str)\n",
    "        new_string = input_str.replace(\"```json\", \"```\")\n",
    "        cleaned_string = new_string.strip().strip('```').strip()\n",
    "            \n",
    "        parsed_dict = json.loads(input_str)\n",
    "        print(parsed_dict)\n",
    "        \n",
    "        json_response = parsed_dict\n",
    "\n",
    "\n",
    "        try:\n",
    "            text_references = json_response['references']\n",
    "        except KeyError:\n",
    "            raise ValueError(\"The response does not contain a 'references' field.\")\n",
    "\n",
    "        try:\n",
    "            question = json_response['question']\n",
    "        except KeyError:\n",
    "            raise ValueError(\"The response does not contain a 'question' field.\")\n",
    "\n",
    "        references = []\n",
    "        for reference in text_references:\n",
    "            if not isinstance(reference, str):\n",
    "                raise ValueError(f\"Expected reference to be of type str, but got {type(reference).__name__}\")\n",
    "            target = rigorous_document_search(corpus, reference)\n",
    "            if target is not None:\n",
    "                reference, start_index, end_index = target\n",
    "                references.append((reference, start_index, end_index))\n",
    "            else:\n",
    "                raise ValueError(f\"No match found in the document for the given reference.\\nReference: {reference}\")\n",
    "\n",
    "        return question, references\n",
    "\n",
    "    def _generate_corpus_questions(self, corpus_id, approx=False, n=5):\n",
    "        with open(corpus_id, 'r') as file:\n",
    "            corpus = file.read()\n",
    "\n",
    "        i = 0\n",
    "        while i < n:\n",
    "            while True:\n",
    "                try:\n",
    "                    print(f\"Trying Query {i}\")\n",
    "                    questions_list = self.synth_questions_df[self.synth_questions_df['corpus_id'] == corpus_id][\n",
    "                        'question'].tolist()\n",
    "                    if approx:\n",
    "                        question, references = self._extract_question_and_approx_references(corpus, 4000,\n",
    "                                                                                            questions_list)\n",
    "                    else:\n",
    "                        question, references = self._extract_question_and_references(corpus, 4000, questions_list)\n",
    "                    if len(references) > 5:\n",
    "                        raise ValueError(\"The number of references exceeds 5.\")\n",
    "\n",
    "                    references = [{'content': ref[0], 'start_index': ref[1], 'end_index': ref[2]} for ref in references]\n",
    "                    new_question = {\n",
    "                        'question': question,\n",
    "                        'references': json.dumps(references),\n",
    "                        'corpus_id': corpus_id\n",
    "                    }\n",
    "\n",
    "                    new_df = pd.DataFrame([new_question])\n",
    "                    self.synth_questions_df = pd.concat([self.synth_questions_df, new_df], ignore_index=True)\n",
    "                    self._save_questions_df()\n",
    "\n",
    "                    break\n",
    "                except (ValueError, json.JSONDecodeError) as e:\n",
    "                    print(f\"Error occurred: {e}\")\n",
    "                    continue\n",
    "            i += 1\n",
    "\n",
    "    def _get_synth_questions_df(self):\n",
    "        if os.path.exists(self.questions_csv_path):\n",
    "            synth_questions_df = pd.read_csv(self.questions_csv_path)\n",
    "        else:\n",
    "            synth_questions_df = pd.DataFrame(columns=['question', 'references', 'corpus_id'])\n",
    "        return synth_questions_df\n",
    "\n",
    "    def generate_queries_and_excerpts(self, approximate_excerpts=False, num_rounds=-1, queries_per_corpus=5):\n",
    "        self.synth_questions_df = self._get_synth_questions_df()\n",
    "\n",
    "        rounds = 0\n",
    "        while num_rounds == -1 or rounds < num_rounds:\n",
    "            for corpus_id in self.corpora_paths:\n",
    "                self._generate_corpus_questions(corpus_id, approx=approximate_excerpts, n=queries_per_corpus)\n",
    "            rounds += 1\n",
    "\n",
    "    def _get_sim(self, target, references):\n",
    "        response = self.client.embeddings.create(\n",
    "            input=[target] + references,\n",
    "            model=\"text-embedding-bge-m3\"\n",
    "        )\n",
    "        nparray1 = np.array(response.data[0].embedding)\n",
    "\n",
    "        full_sim = []\n",
    "        for i in range(1, len(response.data)):\n",
    "            nparray2 = np.array(response.data[i].embedding)\n",
    "            cosine_similarity = np.dot(nparray1, nparray2) / (np.linalg.norm(nparray1) * np.linalg.norm(nparray2))\n",
    "            full_sim.append(cosine_similarity)\n",
    "\n",
    "        return full_sim\n",
    "\n",
    "    def _corpus_filter_poor_highlights(self, corpus_id, synth_questions_df, threshold):\n",
    "        corpus_questions_df = synth_questions_df[synth_questions_df['corpus_id'] == corpus_id]\n",
    "\n",
    "        def edit_row(row):\n",
    "            question = row['question']\n",
    "            references = [ref['content'] for ref in row['references']]\n",
    "            similarity_scores = self._get_sim(question, references)\n",
    "            worst_ref_score = min(similarity_scores)\n",
    "            row['worst_ref_score'] = worst_ref_score\n",
    "            return row\n",
    "\n",
    "        # Apply the function to each row\n",
    "        corpus_questions_df = corpus_questions_df.apply(edit_row, axis=1)\n",
    "\n",
    "        count_before = len(corpus_questions_df)\n",
    "\n",
    "        corpus_questions_df = corpus_questions_df[corpus_questions_df['worst_ref_score'] >= threshold]\n",
    "        corpus_questions_df = corpus_questions_df.drop(columns=['worst_ref_score'])\n",
    "\n",
    "        count_after = len(corpus_questions_df)\n",
    "\n",
    "        print(f\"Corpus: {corpus_id} - Removed {count_before - count_after} .\")\n",
    "\n",
    "        corpus_questions_df['references'] = corpus_questions_df['references'].apply(json.dumps)\n",
    "\n",
    "        full_questions_df = pd.read_csv(self.questions_csv_path)\n",
    "        full_questions_df = full_questions_df[full_questions_df['corpus_id'] != corpus_id]\n",
    "\n",
    "        full_questions_df = pd.concat([full_questions_df, corpus_questions_df], ignore_index=True)\n",
    "        # Drop the columns 'fixed', 'worst_ref_score' and 'diff_score' if they exist\n",
    "        for col in ['fixed', 'worst_ref_score', 'diff_score']:\n",
    "            if col in full_questions_df.columns:\n",
    "                full_questions_df = full_questions_df.drop(columns=col)\n",
    "\n",
    "        full_questions_df.to_csv(self.questions_csv_path, index=False)\n",
    "\n",
    "    def filter_poor_excerpts(self, threshold=0.36, corpora_subset=[]):\n",
    "        if os.path.exists(self.questions_csv_path):\n",
    "            synth_questions_df = pd.read_csv(self.questions_csv_path)\n",
    "            if len(synth_questions_df) > 0:\n",
    "                synth_questions_df['references'] = synth_questions_df['references'].apply(json.loads)\n",
    "                corpus_list = synth_questions_df['corpus_id'].unique().tolist()\n",
    "                if corpora_subset:\n",
    "                    corpus_list = [c for c in corpus_list if c in corpora_subset]\n",
    "                for corpus_id in corpus_list:\n",
    "                    self._corpus_filter_poor_highlights(corpus_id, synth_questions_df, threshold)\n",
    "\n",
    "    def _corpus_filter_duplicates(self, corpus_id, synth_questions_df, threshold):\n",
    "        corpus_questions_df = synth_questions_df[synth_questions_df['corpus_id'] == corpus_id].copy()\n",
    "\n",
    "        count_before = len(corpus_questions_df)\n",
    "\n",
    "        corpus_questions_df.drop_duplicates(subset='question', keep='first', inplace=True)\n",
    "\n",
    "        questions = corpus_questions_df['question'].tolist()\n",
    "\n",
    "        response = self.client.embeddings.create(\n",
    "            input=questions,\n",
    "            model=\"text-embedding-bge-m3\"\n",
    "        )\n",
    "\n",
    "        embeddings_matrix = np.array([data.embedding for data in response.data])\n",
    "\n",
    "        dot_product_matrix = np.dot(embeddings_matrix, embeddings_matrix.T)\n",
    "\n",
    "        # Create a list of tuples containing the index pairs and their similarity\n",
    "        similarity_pairs = [(i, j, dot_product_matrix[i][j]) for i in range(len(dot_product_matrix)) for j in\n",
    "                            range(i + 1, len(dot_product_matrix))]\n",
    "\n",
    "        # Sort the list of tuples based on the similarity in descending order\n",
    "        similarity_pairs.sort(key=lambda x: x[2], reverse=True)\n",
    "\n",
    "        similarity_scores = np.array([x[2] for x in similarity_pairs])\n",
    "\n",
    "        most_similars = (dot_product_matrix - np.eye(dot_product_matrix.shape[0])).max(axis=1)\n",
    "\n",
    "        def filter_vectors(sim_matrix, threshold):\n",
    "            n = sim_matrix.shape[0]  # Number of vectors\n",
    "            remaining = np.ones(n, dtype=bool)  # Initialize all vectors as remaining\n",
    "\n",
    "            for i in range(n):\n",
    "                if remaining[i] == 1:  # Only check for vectors that are still remaining\n",
    "                    for j in range(i + 1, n):\n",
    "                        if remaining[j] == 1 and sim_matrix[i, j] > threshold:\n",
    "                            remaining[j] = 0  # Remove vector j because it's too similar to vector i\n",
    "\n",
    "            return remaining\n",
    "\n",
    "        rows_to_keep = filter_vectors(dot_product_matrix, threshold)\n",
    "\n",
    "        corpus_questions_df = corpus_questions_df[rows_to_keep]\n",
    "\n",
    "        count_after = len(corpus_questions_df)\n",
    "\n",
    "        print(f\"Corpus: {corpus_id} - Removed {count_before - count_after} .\")\n",
    "\n",
    "        corpus_questions_df['references'] = corpus_questions_df['references'].apply(json.dumps)\n",
    "\n",
    "        full_questions_df = pd.read_csv(self.questions_csv_path)\n",
    "        full_questions_df = full_questions_df[full_questions_df['corpus_id'] != corpus_id]\n",
    "\n",
    "        full_questions_df = pd.concat([full_questions_df, corpus_questions_df], ignore_index=True)\n",
    "        # Drop the columns 'fixed', 'worst_ref_score' and 'diff_score' if they exist\n",
    "        for col in ['fixed', 'worst_ref_score', 'diff_score']:\n",
    "            if col in full_questions_df.columns:\n",
    "                full_questions_df = full_questions_df.drop(columns=col)\n",
    "\n",
    "        full_questions_df.to_csv(self.questions_csv_path, index=False)\n",
    "\n",
    "    def filter_duplicates(self, threshold=0.78, corpora_subset=[]):\n",
    "        if os.path.exists(self.questions_csv_path):\n",
    "            synth_questions_df = pd.read_csv(self.questions_csv_path)\n",
    "            if len(synth_questions_df) > 0:\n",
    "                synth_questions_df['references'] = synth_questions_df['references'].apply(json.loads)\n",
    "                corpus_list = synth_questions_df['corpus_id'].unique().tolist()\n",
    "                if corpora_subset:\n",
    "                    corpus_list = [c for c in corpus_list if c in corpora_subset]\n",
    "                for corpus_id in corpus_list:\n",
    "                    self._corpus_filter_duplicates(corpus_id, synth_questions_df, threshold)\n",
    "\n",
    "    def question_ref_filter(self):\n",
    "        self.synth_questions_df = self._get_synth_questions_df()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4484f1ec-b7a6-4f08-9587-646414c7999d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Trying Query 0\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What specific RSS shows no obvious explanation for its lack of activity despite having all critical residues conserved?', 'references': [{'content': 'We have performed an extensive analysis of the functional properties of RSS elements in the context of endogenous recombination signals. To explore the nature of the complex relationships that might exist among different elements and positions in the RSS, we started with the nonfunctional RSS associated with the murine Jβ2.6 pseudogene element of the TCRβ locus (Jβ2.6 RSS). While most such pseudogene elements are flanked by RSSs with crippling mutations (Akira et al. 1987), Jβ2.6 is unique in that the sequence of its flanking RSS suggests no obvious explanation for its complete lack of activity (Figure 1). All of the critical residues are conserved, and each nonconsensus nucleotide in the heptamer and nonamer is represented in at least one other functional RSS in the TCRβ locus (Figure 1).', 'start_chunk': 12, 'end_chunk': 18}]}\n",
      "Trying Query 1\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'Which hypothetical protein is graphically shown in Figure 1E for internal consistency of expression profile measurements?', 'references': [{'content': 'The internal consistency of expression profile measurements for ORFs represented by more than one oligonucleotide sequence is graphically shown in Figure 1E for the hypothetical protein MAL6P1.147, the largest predicted ORF in the genome (31 kb), which is represented by 14 oligonucleotide elements spanning the entire length of the coding sequence.', 'start_chunk': 9, 'end_chunk': 11}]}\n",
      "Trying Query 2\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What are the advantages of using Xenopus egg extracts in the study of the Wnt pathway?', 'references': [{'content': 'There are several unusual advantages to the extract system that contributed to this effort.', 'start_chunk': 4, 'end_chunk': 4}, {'content': 'The Xenopus egg extract is essentially neat cytoplasm; it reproduces the in vivo rate of β-catenin degradation and responds to known regulators as expected from in vivo experiments.', 'start_chunk': 5, 'end_chunk': 5}, {'content': 'Kinetic experiments with high time resolution are possible in this system, since a well-stirred extract is presumably synchronous in ways in which collections of cells may not be.', 'start_chunk': 6, 'end_chunk': 6}, {'content': 'In extracts it is possible to precisely set the level of components by depletion or addition.', 'start_chunk': 7, 'end_chunk': 7}, {'content': 'The direct output of the canonical Wnt pathway is an easily measured cytoplasmic event, the degradation of β-catenin.', 'start_chunk': 9, 'end_chunk': 9}]}\n",
      "Trying Query 3\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What is the result of excluding A+T-rich codon sites from the protein concatenation alignment?', 'references': [{'content': 'Using the resulting alignment (about 30,000 residues), we have reconstructed two trees, one with the NJ method and the polyacid-modified (PAM) matrix; the other with the NJ method and the γ-based method for correcting the rate heterogeneity among sites and bootstrap. The trees obtained (data not shown) were identical and gave strong support to the grouping of Buchnera and Wigglesworthia and to their position as the sister-group of enteric bacteria (Escherichia, Salmonella, and Yersinia). Thus, this grouping is probably not an artifact of the biased composition of the endosymbiont genomes.', 'start_chunk': 17, 'end_chunk': 23}]}\n",
      "Trying Query 4\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What is the updated date for the content described in the text?', 'references': [{'content': 'Last Updated: 2021-01-05 08:21:04', 'start_chunk': 28, 'end_chunk': 29}]}\n",
      "Trying Query 5\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What are the two endosymbionts mentioned in the text and what is their relationship based on the phylogenetic framework?', 'references': [{'content': 'An example of the kind of biological inference that can be built upon a well-supported phylogeny is provided by the two endosymbionts included in our set. Wigglesworthia and Buchnera have sometimes been considered as closely related and sometimes not, based on relatively weak phylogenetic evidence provided by the SSU rRNA alone. Our confirmation of their close relationship raises the question of whether their common ancestor was an endosymbiont with a reduced genome or a free-living bacterium (perhaps one with a host-associated lifestyle that promoted formation of intimate symbiosis).', 'start_chunk': 1, 'end_chunk': 7}]}\n",
      "Trying Query 6\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'Which experiment demonstrated that brefeldin A-sensitive guanine nucleotide-exchange protein is involved in ADP-ribosylation factor for a specific organism?', 'references': [{'content': 'Morinaga N Moss J Vaughan M Cloning and expression of a cDNA encoding a bovine brain brefeldin A-sensitive guanine nucleotide-exchange protein for ADP-ribosylation factor Proc Natl Acad Sci U S A 1997 94 12926 12931 9371777', 'start_chunk': 1, 'end_chunk': 2}]}\n",
      "Trying Query 7\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'Who manages Creative Commons?', 'references': [{'content': 'Creative Commons, the organization I help manage, comes in.', 'start_chunk': 0, 'end_chunk': 0}]}\n",
      "Trying Query 8\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'Which ARNO mutants did not co-immunoprecipitate with the insulin receptor?', 'references': [{'content': 'However, none of the deletion mutants, including ΔPH-ARNO, PH-ARNO, ΔCC-ARNO, and CC-ARNO, as well as a site-directed mutant R280D-ARNO, was found co-immunoprecipitated with the insulin receptor.', 'start_chunk': 7, 'end_chunk': 9}]}\n",
      "Trying Query 9\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'At what time does the peak expression of ribonucleotide reductase occur?', 'references': [{'content': 'mRNA transcripts for enzymes that convert ribonucleotides into deoxyribonucleotides, including DHFR-TS and both subunits of ribonucleotide reductase, were induced approximately at 10 hpi, peaking at approximately 32 hpi.', 'start_chunk': 2, 'end_chunk': 5}]}\n",
      "Trying Query 10\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'How do consensus spacer sequences affect the formation of a stable 12-SC complex?', 'references': [{'content': 'If strong interactions with the nonamer can form, then the presence of a consensus spacer may allow additional favorable contacts to be established, not only in the spacer itself, but even farther away, in the heptamer.', 'start_chunk': 1, 'end_chunk': 1}, {'content': \"By contrast, an unfavorable spacer (e.g., the Jβ2.6 RSS spacer or Sac) may structurally 'insulate' protein–DNA contacts seen in the nonamer, such that potential heptamer contact surfaces that could otherwise contribute to overall 12-SC stability remain hidden.\", 'start_chunk': 2, 'end_chunk': 3}, {'content': 'This may explain why a consensus heptamer, in the absence of a good nonamer, is unable to promote formation of a stable 12-SC complex.', 'start_chunk': 5, 'end_chunk': 6}]}\n",
      "Trying Query 11\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'Which RSS associated with the murine Jβ2.6 pseudogene element is unique for its lack of activity despite having all critical residues conserved?', 'references': [{'content': 'We have performed an extensive analysis of the functional properties of RSS elements in the context of endogenous recombination signals. To explore the nature of the complex relationships that might exist among different elements and positions in the RSS, we started with the nonfunctional RSS associated with the murine Jβ2.6 pseudogene element of the TCRβ locus. While most such pseudogene elements are flanked by RSSs with crippling mutations (Akira et al. 1987), Jβ2.6 is unique in that the sequence of its flanking RSS suggests no obvious explanation for its complete lack of activity (Figure 1).', 'start_chunk': 31, 'end_chunk': 34}]}\n",
      "Trying Query 12\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What are the three components that make up each recombination signal sequence (RSS) during V(D)J recombination?', 'references': [{'content': 'Sequence alignments of RSSs suggested that each signal can be dissected into three components: a conserved heptamer (consensus: 5′-CACAGTG) and a conserved nonamer (consensus: 5′-ACAAAAACC), separated by a poorly conserved spacer of either 12 ± 1 or 23 ± 1 bp (Tonegawa 1983; Akira et al. 1987; Ramsden et al. 1994).', 'start_chunk': 29, 'end_chunk': 32}, {'content': 'The heptamer is the site of DNA cleavage (Roth et al. 1992), while the nonamer provides a major binding surface for RAG1 (Difilippantonio et al. 1996; Spanopoulou et al. 1996; Nagawa et al. 1998; Swanson and Desiderio 1998).', 'start_chunk': 33, 'end_chunk': 34}]}\n",
      "Trying Query 13\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'Which protein is secreted by ventral lateral neurons to sustain normal cycling of circadian rhythms in constant darkness?', 'references': [{'content': 'The researchers also demonstrate that the persistence of normal cycling during constant darkness depends on a protein (called PDF) secreted by the ventral lateral cells.', 'start_chunk': 37, 'end_chunk': 39}]}\n",
      "Trying Query 14\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What criteria were used to retain quality spots in the microarray hybridizations?', 'references': [{'content': 'Quality spots were retained based on the following criteria. The log2(Cy5/Cy3) ratio for array features that were unflagged and had a sum of median intensities greater than the local background plus two times the standard deviation of the background were extracted from the database for further analysis.', 'start_chunk': 17, 'end_chunk': 19}]}\n",
      "Trying Query 15\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What is the genetic distance between the Borneo haplotype BD and the closest β-haplotypes?', 'references': [{'content': 'Uncorrected p distances between the Borneo haplotype and other β-haplotypes ranged from 0.012 (haplotypes BQ, BP, BO, BS, BU) to 0.020 (haplotype BE), with a mean of 0.014.', 'start_chunk': 5, 'end_chunk': 6}]}\n",
      "Trying Query 16\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'Which organizations supported RH in this study?', 'references': [{'content': 'Itute of General Medical Sciences and the Harvard-Armenise Foundation for supplemental support (for RH). RH is also supported by the German Research Foundation (grant He 2049/2-1).', 'start_chunk': 0, 'end_chunk': 2}]}\n",
      "Trying Query 17\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What are the factors driving the division between the United States and European Union regarding the use of GM crops?', 'references': [{'content': 'Opinions differ on this, however, and seem to follow the United States–European Union divide over the use of GM crops.', 'start_chunk': 3, 'end_chunk': 4}, {'content': 'To two academicians that kindled the biotech revolution, the real GM risks lie in how science is misinterpreted and misused.', 'start_chunk': 17, 'end_chunk': 18}, {'content': 'Richard Jefferson, chairman and chief executive officer of the Center for the Application of Molecular Biology to International Agriculture in Australia, fears that innovation has been stifled by corporate short-sightedness.', 'start_chunk': 22, 'end_chunk': 24}]}\n",
      "Trying Query 18\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What are the determinants of RSS specificity and function during V(D)J recombination?', 'references': [{'content': 'Only the first three nucleotides of the heptamer and the fifth and sixth positions of the nonamer show almost perfect conservation (Ramsden et al. 1994) and are therefore thought to be the major determinants of RSS specificity and function. Mutations in any of these five “critical” nucleotides, alone or in combination, essentially abolish recombination (Tonegawa 1983; Akira etal. 1987; Hesse et al. 1989).', 'start_chunk': 15, 'end_chunk': 18}]}\n",
      "Trying Query 19\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What is the observed phase change in Pdf01 mutant flies compared to wild-type flies after 4 days of DD?', 'references': [{'content': 'Despite the gradual fading of locomotor activity rhythms in DD, a significant fraction of Pdf01 mutant flies is still weakly rhythmic after 4 d of DD (Renn et al. 1999). By tracking their locomotor activity phases, we observed that most of them had accumulated an approximately 4-hour phase advance relative to wild-type flies by the fourth day in DD.', 'start_chunk': 11, 'end_chunk': 14}]}\n",
      "Trying Query 20\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'How does APC-mediated degradation of axin affect β-catenin degradation with changing APC concentration?', 'references': [{'content': 'APC-mediated degradation of axin (“with regulatory loop” where Equation [5] is applied) or the absence of such an effect (where the linear rate equation ν15 = k15axin is applied). With APC-mediated axin degradation, β-catenin degradation is affected very little by changes in the concentration of APC (25% decrease with a 2-fold increase in APC concentration).', 'start_chunk': 0, 'end_chunk': 2}, {'content': 'Decreasing the concentration of APC inhibits the degradation of axin, thereby promoting the formation of the degradation complex. As shown in Figure 5, in the absence of the regulatory loop, axin degradation is APC independent, homeostasis is lost, and β-catenin levels are greatly upregulated with decreasing APC concentrations.', 'start_chunk': 5, 'end_chunk': 8}]}\n",
      "Trying Query 21\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': \"What methods did the researchers use to determine the evolutionary path of Borneo's elephants?\", 'references': [{'content': \"Applying DNA analysis and dating techniques to investigate the elephants' evolutionary path, researchers from the United States, Ind\", 'start_chunk': 41, 'end_chunk': 42}]}\n",
      "Trying Query 22\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': \"Who conducted the DNA analysis to determine the evolutionary path of Borneo's elephants, as reported by Fernando et al.?\", 'references': [{'content': \"Applying DNA analysis and dating techniques to investigate the elephants' evolutionary path, researchers from the United States, India, and Malaysia, led by Don Melnick of the Center for Environmental Research and Conservation at Columbia, demonstrate that Borneo's elephants are not recent arrivals.\", 'start_chunk': 25, 'end_chunk': 27}]}\n",
      "Trying Query 23\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'How many loci required for viability were detected using rrf-3 compared to N2?', 'references': [{'content': 'We next asked whether using the rrf-3 strain improved general phenotype detection or whether certain types of phenotypes were particularly increased compared to the N2 screens by Fraser et al. (2000) and Kamath et al. (2003). To do this, we analysed the detection rate of different types of Chromosome I loci. First, we looked at a set of 23 loci with nonlethal postembryonic mutant phenotypes. Using rrf-3, we reproducibly detected the published phenotype for 11 of these compared to only two for N2. Of 50 loci required for viability (essential genes), we detected 31 using rrf-3, compared to 33 for N2.', 'start_chunk': 8, 'end_chunk': 14}]}\n",
      "Trying Query 24\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'Which study explores the role of SERA genes in the erythrocytic cycle of Plasmodium falciparum?', 'references': [{'content': '31206 Miller SK Good RT Drew DR Delorenzi M Sanders PR A subset of Plasmodium falciparum SERA genes are expressed and appear to play an important role in the erythrocytic cycle J Biol Chem 2002 277 47524 47532 12228245', 'start_chunk': 0, 'end_chunk': 1}]}\n",
      "Trying Query 25\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': \"What is the role of Creative Commons licenses in the Massachusetts Institute of Technology's Open Courseware project?\", 'references': [{'content': \"The Massachusetts Institute of Technology's Open Courseware project publishes materials from its university courses under a version of the licenses, inviting students and educators from around the world to reuse them royalty-free.\", 'start_chunk': 15, 'end_chunk': 16}]}\n",
      "Trying Query 26\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What specific conditions were used to simulate the addition of 1 μM activated Dsh in the degradation literature?', 'references': [{'content': 'Curve c: addition of 1 μM activated Dsh (deactivation of Dsh was neglected, k2 = 0);', 'start_chunk': 26, 'end_chunk': 28}, {'content': 'Curve c′: 1 μM Dsh.', 'start_chunk': 32, 'end_chunk': 33}]}\n",
      "Trying Query 27\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What stage of the P. falciparum life cycle is described as having a 48-hour intraerythrocytic developmental cycle (IDC) with merozoite invasion and the formation of the parasitophorous vacuole?', 'references': [{'content': 'The 48-h P. falciparum IDC (Figure 1A) initiates with merozoite invasion of red blood cells (RBCs) and is followed by the formation of the parasitophorous vacuole (PV) during the ring stage.', 'start_chunk': 24, 'end_chunk': 26}]}\n",
      "Trying Query 28\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What was the primary function of the rrf-3 mutant in the RNAi library feeding experiments?', 'references': [{'content': 'To increase the number of genes for which a loss-of-function phenotype can be detected, we undertook a genome-wide RNAi screen using the rrf-3 mutant strain, which we found to be hypersensitive to RNAi.', 'start_chunk': 3, 'end_chunk': 5}, {'content': 'Feeding of the RNAi library to rrf-3 mutants resulted in additional loss-of-function phenotypes for 393 genes, increasing the number of genes with a phenotype by 23%.', 'start_chunk': 5, 'end_chunk': 7}, {'content': 'We have already shown that mutation of rrf-3, a putative RNA-directed RNA polymerase (RdRP), resulted in increased sensitivity to RNAi.', 'start_chunk': 31, 'end_chunk': 32}, {'content': 'The rrf-3 strain, mutated in the fourth RdRP homolog, shows an opposite response to dsRNA; this mutant has increased sensitivity to RNAi.', 'start_chunk': 36, 'end_chunk': 37}, {'content': 'A more detailed study of RNAi sensitivity of rrf-3 mutants using a set of 80 genes showed that rrf-3 is generally more sensitive to RNAi than wild-type worms.', 'start_chunk': 37, 'end_chunk': 38}]}\n",
      "Trying Query 29\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': \"Who are the authors of 'Identification and utilization of arbitrary correlations in models of recombination signal sequences'?\", 'references': [{'content': 'Cowell LG  Davila M  Kepler TB  Kelsoe G   Identification and utilization of arbitrary correlations in models of recombination signal sequences Genome Biol 2002 3 research0072.1', 'start_chunk': 5, 'end_chunk': 6}]}\n",
      "Trying Query 30\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'At what specific points during a circadian day were brains taken for in situ hybridization in the fourth day of DD?', 'references': [{'content': 'Brains taken in the fourth day of DD. Six timepoints were taken throughout the circadian day. The sequence of panels from left to right is CT2, 6, 10, 14, 18, and 20, respectively.', 'start_chunk': 20, 'end_chunk': 21}]}\n",
      "Trying Query 31\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'How many genes are induced during the ring to early trophozoite stage in P. falciparum IDC?', 'references': [{'content': 'Following invasion, approximately 950 ORFs are induced during the ring and early trophozoite stage, including genes associated with the cytoplasmic transcriptional and translational machinery, glycolysis and ribonucleotide biosynthesis.', 'start_chunk': 17, 'end_chunk': 19}]}\n",
      "Trying Query 32\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'Where in the brain was the strong PDF peptide signal detected in flies labeled with LNvs?', 'references': [{'content': 'The brain is from flies with labeled LNvs (y w,UAS–mCD8iGFP;pdf–GAL4). Numerous cells at the periphery of the medulla have the vast majority of the bound PDF peptide signal within the brain.', 'start_chunk': 10, 'end_chunk': 12}]}\n",
      "Trying Query 33\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What percentage of the Plasmodium falciparum genome is devoted to a system of antigenic determinants for host immune evasion?', 'references': [{'content': 'The Plasmodium parasite devotes 3.9% of its genome to a complex system of antigenic determinants essential for host immune evasion during a single developmental stage (Gardner et al. 2002).', 'start_chunk': 2, 'end_chunk': 3}]}\n",
      "Trying Query 34\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun.\", 'question': 'How do researchers examine discordance in topology derived from gene alignments?', 'references': [{'content': 'topology and two gene alignments that rejected that topology. For each case, we observed which taxon showed the most evident discordance in the topology derived from the exceptional gene alignment.', 'start_chunk': 0, 'end_chunk': 1}, {'content': 'We then removed the corresponding sequence from the alignment and the corresponding taxon from the widely favored topologies. Using an SH test, we determined whether the alignment continued to show significant conflict with the favored topologies.', 'start_chunk': 2, 'end_chunk': 4}]}\n",
      "Trying Query 35\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What strongly suggests that the Borneo elephant population was not derived from another population in the recent past according to microsatellite data?', 'references': [{'content': 'In the assignment tests, all five Borneo genotypes, which included free-ranging as well as captive animals, were assigned to Borneo with significantly higher likelihoods than to other populations and with extremely low likelihoods to the putative source populations.', 'start_chunk': 28, 'end_chunk': 30}, {'content': 'An introduced population may be highly divergent from the source population in terms of F statistics (Williams et al. 2002) due to allelic loss from founder-events. However, the probability of loss for a particular allele is inversely proportional to its frequency in the founder and hence the source population. Thus, genotypes in an introduced population would retain a high likelihood of assignment to the source population, enabling its identification from among a number of candidate populations.', 'start_chunk': 31, 'end_chunk': 34}, {'content': 'Therefore, the assignment tests strongly suggest that the Borneo elephants were not derived from another population in the recent past.', 'start_chunk': 35, 'end_chunk': 37}]}\n",
      "Trying Query 36\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'How did the presence of insulin influence ARF1-GFP translocation to the plasma membrane in HeLa cells?', 'references': [{'content': 'Ten minutes after the insulin stimulation, most of the ARF1-GFP was found on the plasma membrane.', 'start_chunk': 0, 'end_chunk': 1}, {'content': 'Since the endogenous levels of ARNO in HeLa cells were so low that the protein could not be detected in Western blots, it is reasonable to assume that under physiological conditions only a very small fraction of ARF1 translocates to the plasma membrane in response to extracellular agonists.', 'start_chunk': 4, 'end_chunk': 6}, {'content': 'Receptors that were immunoprecipitated in the presence of insulin were associated with an ARF-GEF activity, and that the ARF-GEF activity that was co-immunoprecipitated with the insulin receptor was significantly increased in the cells that had been transiently transfected with myc-ARNO.', 'start_chunk': 23, 'end_chunk': 25}]}\n",
      "Trying Query 37\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What is the ratio of dissociation constants for β-catenin binding to APC and APC*/axin*/GSK3β?', 'references': [{'content': 'the ratio K17/K8 = 10 of the dissociation constants characterizing the binding of β-catenin to APC and APC*/axin*/GSK3β, respectively, is based on previous experimental results demonstrating that β-catenin has a 10-fold lower affinity for nonphosphorylated compared to phosphorylated APC (Salic et al. 2000)', 'start_chunk': 3, 'end_chunk': 5}]}\n",
      "Trying Query 38\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What effect does increasing the concentration of Dsh have on β-catenin degradation according to the experimental results?', 'references': [{'content': 'As expected, β-catenin degradation is inhibited by increasing Dsh concentration and stimulated by increasing the concentration of either axin or GSK3β.', 'start_chunk': 35, 'end_chunk': 36}, {'content': '(C) Axin increases the rate of β-catenin degradation even in the absence of added Dsh.', 'start_chunk': 24, 'end_chunk': 26}, {'content': '(D) Stimulation of β-catenin degradation by GSK3β is detected only at high concentrations of Dsh. No effects of GSK3β on β-catenin degradation can be detected at less than 30 nM added Dsh.', 'start_chunk': 26, 'end_chunk': 28}, {'content': 'These curves represent β-catenin half-lives for various concentrations of axin (Figure 4C) and GSK3β (Figure 4D) with varying concentrations of Dsh.', 'start_chunk': 31, 'end_chunk': 33}, {'content': 'The results are qualitatively similar to those predicted by the model.', 'start_chunk': 33, 'end_chunk': 34}]}\n",
      "Trying Query 39\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What is the half-life of β-catenin when nonaxin-dependent processes contribute 10% to its total degradation?', 'references': [{'content': 'Experimentally, we have measured only a 1.5% contribution to total β-catenin degradation such that the half-life of β-catenin is 45 h when the axin-dependent processes are inhibited. If in some situations the nonaxin-dependent degradation contributed 10% to the flux, the half-life would be 6.3 h (k 13 = 1.83 · 10−3 min−1).', 'start_chunk': 4, 'end_chunk': 7}]}\n",
      "Trying Query 40\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What is the range of new merozoites formed through the replication and division process?', 'references': [{'content': 'y replicating and dividing to form up to 32 new merozoites.', 'start_chunk': 0, 'end_chunk': 0}]}\n",
      "Trying Query 41\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What percentage of the Plasmodium genome is devoted to antigenic determinants for host immune evasion during a single developmental stage?', 'references': [{'content': 'The Plasmodium parasite devotes 3.9% of its genome to a complex system of antigenic determinants essential for host immune evasion during a single developmental stage.', 'start_chunk': 19, 'end_chunk': 21}]}\n",
      "Trying Query 42\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What is the license type and citation for the PLoS Biology article?', 'references': [{'content': 'License: CC BY', 'start_chunk': 33, 'end_chunk': 33}, {'content': 'Citation: PLoS Biol. 2003 Oct 15; 1(1):e19', 'start_chunk': 34, 'end_chunk': 34}]}\n",
      "Trying Query 43\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'Which proteases are implicated in the merozoite invasion stage according to Greenbaum et al. 2002 and potentially localized apically in dense granules?', 'references': [{'content': 'FP-1 has recently been implicated in merozoite invasion and has been localized to the interior of the PV (Greenbaum et al. 2002).', 'start_chunk': 13, 'end_chunk': 15}, {'content': 'The expression of PM V and the newly identified FP-2 homologue during this stage suggests they are also important in the trophozoite stage.', 'start_chunk': 19, 'end_chunk': 21}, {'content': 'PfSUB1 and PfSUB2 are believed to be involved in merozoite invasion and have been localized apically in the dense granules.', 'start_chunk': 31, 'end_chunk': 32}]}\n",
      "Trying Query 44\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'How does the substitution of individual consensus elements H, Sc, and N affect RAG protein binding efficiency?', 'references': [{'content': 'Substitution of the individual consensus elements H, Sc, and N, however, led to surprising results. While the consensus nonamer (N) sequence, as expected, increases the level of binding (up to that of Jβ2.2), the consensus spacer (Sc) alone has no effect on binding at all, and the consensus heptamer (H) consistently reduces the level of binding.', 'start_chunk': 26, 'end_chunk': 29}, {'content': 'The consensus spacer boosts binding only in the context of a consensus nonamer (the ratios of Sc–N:N and H–Sc–N:H–N are greater than H–Sc:H), and the consensus heptamer contributes significantly to RAG–RSS interactions in this assay only when both spacer and nonamer are consensus sequences (H–Sc–N:Sc–N > H–N:N or H:Jβ2.6 RSS).', 'start_chunk': 29, 'end_chunk': 32}]}\n",
      "Trying Query 45\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What criteria were used to eliminate nonhomologous and paralogous sequences?', 'references': [{'content': 'We applied a stringent criterion for eliminating nonhomologous sequences and paralogous sequences, since both are likely to lead to false conclusions regarding the organismal phylogeny and frequency of LGT.', 'start_chunk': 22, 'end_chunk': 23}, {'content': 'Instead, we used a cutoff for the degree of similarity as reflected in the BLASTP bit scores (Altschul et al. 1997). The bit score is dependent upon the scoring system (substitution matrix and gap costs) employed and takes into account both the degree of similarity and the length of the alignment between the query and the match sequences.', 'start_chunk': 27, 'end_chunk': 29}]}\n",
      "Trying Query 46\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'Which gene was identified in the cloning of genetic mutants and what mutation does it have?', 'references': [{'content': 'The identification of mutations in unc-108 encoding the homolog of the small GTPase Rab2 is of particular interest.', 'start_chunk': 36, 'end_chunk': 37}, {'content': 'dpy-6(e2762) has a deletion that removes the first six amino acid residues (aa) of the eighth exon and part of the seventh intron.', 'start_chunk': 32, 'end_chunk': 33}, {'content': 'Multiple mutations in dpy-6(f11) (5′-tcgAaaa[G/T]tt[C/A]aaccccacgccaact[G/T]cc); the AAA→AAAA mutation at position 2792 bp of the F16F9.2 coding sequence causes a frameshift that results in a premature stop in the fifth exon.', 'start_chunk': 33, 'end_chunk': 36}]}\n",
      "Trying Query 47\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'Which recent study involving Drosophila suggests that LNvs control the prothoracic gland clock via PDF signaling?', 'references': [{'content': 'A very recent study of the Drosophila prothoracic gland (PG) clock and eclosion rhythms suggests that the LNvs also control the PG clock via PDF signaling (Myers et al. 2003).', 'start_chunk': 1, 'end_chunk': 2}]}\n",
      "Trying Query 48\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'What role does insulin play in the translocation of ARNO and ARF to the plasma membrane?', 'references': [{'content': 'Our studies showed that insulin promoted the translocation of myc-tagged ARNO constructs to the plasma membrane.', 'start_chunk': 9, 'end_chunk': 10}, {'content': 'The linkage between ARNO translocation to specific subcellular fractions and ARF activation was studied using myc-tagged ARNO and ARF-GFP constructs in two different cell types. Our data showed conclusively that insulin promoted the co-localization of wild type myc-ARNO and ARF1-GFP on the surface of HIRcB and HeLa cells.', 'start_chunk': 18, 'end_chunk': 20}, {'content': 'Interestingly, insulin, acting through ARNO, promoted the translocation of ARF1-GFP to the plasma membrane.', 'start_chunk': 20, 'end_chunk': 21}]}\n",
      "Trying Query 49\n",
      "{'oath': \"I will not use the word 'and' in the question unless it is part of a proper noun. I will also make sure the question is concise.\", 'question': 'Which publications relate to the sequence evolution in bacterial endosymbionts with extreme base composition?', 'references': [{'content': 'Clark MA  Moran NA  Baumann P   Sequence evolution in bacterial endosymbionts having extreme base composition Mol Biol Evol 1999 16 1586 1598 10555290', 'start_chunk': 12, 'end_chunk': 14}]}\n"
     ]
    }
   ],
   "source": [
    "# Specify the corpora paths, you can have multiple but we'll just use our one file\n",
    "corpora_paths = [\n",
    "    'chunking_evaluation/pubmed_data/pubmed.txt',\n",
    "]\n",
    "\n",
    "csv_path = 'chunking_evaluation/pubmed_results_phi-4.csv'\n",
    "\n",
    "# Initialize the evaluation\n",
    "synthetic_pipeline = SyntheticEvaluation(corpora_paths, csv_path)\n",
    "synthetic_pipeline.generate_queries_and_excerpts(approximate_excerpts=True,\n",
    "                                         num_rounds=1,\n",
    "                                         queries_per_corpus=50)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8551d3ea-6da6-4765-af45-fd5d48fe3e9e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "2025",
   "language": "python",
   "name": "papers_2025"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
